File: src/09.tree/17.671.go
package tree

func FindSecondMinimumValue1(root *TreeNode) int {
	var dfs func(nodt *TreeNode, cur int)
	ans := -1

	dfs = func(node *TreeNode, cur int) {
		if node == nil {
			return
		}

		if node.Val != cur {
			if ans == -1 {
				ans = node.Val
			} else {
				if ans > node.Val {
					ans = node.Val
				}
			}
			return
		}

		dfs(node.Left, cur)
		dfs(node.Right, cur)
	}

	dfs(root, root.Val)

	return ans
}

func FindSecondMinimumValue(root *TreeNode) int {
	var dfs func(*TreeNode)
	ans := -1
	rootVal := root.Val
	dfs = func(node *TreeNode) {
		if node == nil || ans != -1 && node.Val >= ans {
			return
		}

		if node.Val > rootVal {
			ans = node.Val
		}

		dfs(node.Left)
		dfs(node.Right)
	}

	dfs(root)

	return ans
}

File: src/09.tree/17.671.ts
import { TreeNode } from "./01.100";

export function findSecondMinimumValue(root: TreeNode | null): number {
    let ans = -1;

    const dfs = (node: TreeNode | null, cur: number) => {
        if (!node) {
            return;
        }

        if (node.val !== cur) {
            if (ans === -1) {
                ans = node.val;
            } else {
                ans = Math.min(ans, node.val);
            }
            return;
        }

        dfs(node.left, cur);
        dfs(node.right, cur);
    };

    dfs(root, root.val);

    return ans;
}
